File: Stable_Matching/stable_match.py
# stable matching problem
## 1. need to define the preferences of men and women
### men | women
###  A  |   X
###  B  |   Y
###     |   Z


men_pref = {    'Mike':   ['Rachel', 'Katrina', 'Donna', 'Sheila'],
    'Harvey': ['Donna', 'Katrina', 'Rachel', 'Sheila'],
    'Louis':  ['Sheila', 'Donna', 'Katrina', 'Rachel'],
    'Logan':  ['Rachel', 'Katrina', 'Donna', 'Sheila']}
women_pref = {    'Rachel':  ['Mike', 'Logan', 'Harvey', 'Louis'],
    'Donna':   ['Harvey', 'Louis', 'Mike', 'Logan'],
    'Katrina': ['Mike', 'Harvey', 'Louis', 'Logan'],
    'Sheila':  ['Louis', 'Logan', 'Harvey', 'Mike']}
## lists of men and women who are free
men_free = list(men_pref.keys())
women_free = list(women_pref.keys())

## create a dictionary to store the matches
men_match = {man : '' for man in men_pref.keys()}
key_list = list(men_pref.keys())

while len(men_free) > 0: # loop until there are no men free
    for man in key_list: # loop through the men list
        for woman in men_pref[man]: # loop through each man's preference list
            if woman not in list(men_match.values()): # if the woman is not already matched; case 1: both man and woman are unmatched
                men_match[man] = woman # match the woman with the corresponding man
                men_free.remove(man) # remove the matched man from the free men list
                print(f"{man} is matched with {woman}. {man} is no longer free. We have {len(men_free)} men left to match")
                break # get out of the loop since we have matched the couple
            elif woman in list(men_match.values()): # if the woman is already matched; case 2: woman is matched with another man
                current_man = list(men_match.keys())[list(men_match.values()).index(woman)] # get the current man who is matched with the woman
                w_list = women_pref.get(woman) # get the woman's preference list ## get(key) <- bring the value of the dictionary key
                if w_list.index(man) < w_list.index(current_man): # if the currently matched man is less preferred than the new man; case 2a: woman prefers the new
                    men_match[man] = woman
                    men_free.remove(man) # remove the matched
                    men_match[current_man] = '' # set the current man to be unmatched
                    men_free.append(current_man) # add the current man back to the free
                    print(f"{woman} used to be matched with {current_man} but now is matched with {man}. {current_man} is now free.")
                    
                

File: Stable_Matching/stable_match.java
package Stable_Matching;
/*
Stable Matching - Gale-Shapley Algorithm in Java
Matching between Hospital and Student

runtime: O(n^2)
both hospitals and students have IDs: 1,2,...,n


0. Initialization & Variable Setup
INPUT:
h_prefs: hospital preference to student; 2D int array
    h_prefs[h,i]: preference of hospital h for student i; ith favorite student on h's preference
s_prefs: student preference to hospital; 2D int array
    s_prefs[s,j]: preference of student s for hospital j; jth favorite hospital on s's preference

OUTPUT:
job; where job[r]=h if student r matched to hospital h (array of size n)

Tracking engagements:
a queue of `free_hospitals` with operation: empty(), enqueue(), dequeue()

two arrays `job` and `resident` that store who is "engaged" to whom
    - if hospital h is engaged to r, job[r] = h, resident[h]=r
    - entries for unmatched students/hospitals: -1

Hospital proposing: two arrays `hospital_prefs` and `count` (length n)
    - hospital_prefs[h]: list of students preferred by hospital h
    - count[h]: number of proposals made by hospital h
*/
import MyLibrary.MyQueue.*;
import java.util.Arrays;

class stable_match{
    // Fields
    public int n = 5; // number of hospitals and students fixed to 5 for simulation
    private Integer[] job = new Integer[n]; // job array length n
    private Integer[] resident = new Integer[n]; // resident array length n
    private Integer[] count = new Integer[n]; // hospital proposal count of length n

    private void initialize(){
        for (int i=0; i<n;i++){
            this.job[i] = -1;
            this.resident[i] = -1;
            this.count[i] = 0;
        }
    }

    public Integer[] stable_matching_hos_propose(Integer[][] h_prefs, Integer[][] s_prefs){
        initialize();
        // Initialize free hospitals queue
        MyQueue<Integer> free_hospitals = new MyQueueArray<Integer>(n);
        free_hospitals.foritm(
            (hos) -> {
                free_hospitals.enque$exn(s_prefs[0][hos]);
            }
        );

        // Main loop: loop until all hospitals are matched
        while (!free_hospitals.isEmpty()){
            Integer h = free_hospitals.deque$exn(); // pick a free hospital
            Integer r = h_prefs[h][count[h]]; // get the next student to propose to
            count[h]++; // increment the proposal count for hospital h

            // 1st case: if student r is free
            if (job[r].equals(-1)){
                job[r] = h; // student r is assigned to hospital h
                resident[h]=r; // hospital h hired student r as resident
            } else if (s_prefs[r][h] < s_prefs[resident[h]][h]){ // 2nd case: if student r is not free AND prefers hospital h over the current hospital h' (I don't have to reiterate job[r]==-1 since else if already checks this off)
                Integer h_prime = job[r]; // current hospital that student r is matched to
                free_hospitals.enque$exn(h_prime); // put h'(current hospital) back into the free hospitals queue
                job[r] = h; // student r is reassigned to new hospital h
                resident[h] = r; // new hospital h hired student r as resident
            } else {
                free_hospitals.enque$exn(h); // put hospital h back into the free hospitals queue
            }
        };
        return job; // return the final matching of students to hospitals
    }

    public static void main(String[] args){
        // preference list
        Integer[][] h_prefs = {
            {0, 1, 2, 3, 4},   // Hospital 0
            {1, 0, 2, 4, 3},   // Hospital 1
            {2, 3, 1, 0, 4},   // Hospital 2
            {3, 2, 4, 1, 0},   // Hospital 3
            {4, 3, 2, 1, 0}    // Hospital 4
        };

        Integer[][] s_prefs = {
            {1, 0, 2, 3, 4},   // Student 0
            {0, 1, 3, 2, 4},   // Student 1
            {2, 3, 1, 0, 4},   // Student 2
            {3, 2, 4, 1, 0},   // Student 3
            {4, 3, 2, 0, 1}    // Student 4
        };
        /*
        expected output: [0,1,2,3,4] */
        stable_match match = new stable_match();
        
        System.out.println("Stable Matching: " + Arrays.toString(match.stable_matching_hos_propose(h_prefs, s_prefs)));
    }
}